vm/systemSmartContracts/freezeAccount.go

```diff
@@ -0,0 +1,284 @@
+		fa.systemEI.AddReturnMessage(err.Error())
+		return vmcommon.OutOfGas
+	}
+	if !fa.isAddressValid(args.Arguments[0]) {
+		fa.systemEI.AddReturnMessage("invalid address")
+		return vmcommon.UserError
+	}
+	if bytes.Equal(args.CallerAddr, args.Arguments[0]) {
+		fa.systemEI.AddReturnMessage("cannot set own address as guardian")
+		return vmcommon.UserError
+	}
+
+	guardians, err := fa.guardians(args.CallerAddr)
+	if err != nil {
+		fa.systemEI.AddReturnMessage(err.Error())
+		return vmcommon.ExecutionFailed
+	}
+	// Case 1
+	if len(guardians.Data) == 0 {
+		err = fa.addGuardian(args.CallerAddr, args.Arguments[0], guardians)
+		if err != nil {
+			fa.systemEI.AddReturnMessage(err.Error())
+			return vmcommon.ExecutionFailed
+		}
+		return vmcommon.Ok
+		// Case 2
+	} else if len(guardians.Data) == 1 && fa.pending(guardians.Data[0]) {
+		fa.systemEI.AddReturnMessage(fmt.Sprintf("owner already has one guardian pending: %s",
+			fa.pubKeyConverter.Encode(guardians.Data[0].Address)))
+		return vmcommon.UserError
+		// Case 3
+	} else if len(guardians.Data) == 1 && !fa.pending(guardians.Data[0]) {
+		err = fa.addGuardian(args.CallerAddr, args.Arguments[0], guardians)
+		if err != nil {
+			fa.systemEI.AddReturnMessage(err.Error())
+			return vmcommon.ExecutionFailed
+		}
+		return vmcommon.Ok
+		// Case 4
+	} else if len(guardians.Data) == 2 && fa.pending(guardians.Data[1]) {
+		fa.systemEI.AddReturnMessage(fmt.Sprintf("owner already has one guardian pending: %s",
+			fa.pubKeyConverter.Encode(guardians.Data[1].Address)))
+		return vmcommon.UserError
+		// Case 5
+	} else if len(guardians.Data) == 2 && !fa.pending(guardians.Data[1]) {
+		guardians.Data = guardians.Data[1:] // remove oldest guardian
+		err = fa.addGuardian(args.CallerAddr, args.Arguments[0], guardians)
+		if err != nil {
+			fa.systemEI.AddReturnMessage(err.Error())
+			return vmcommon.ExecutionFailed
+		}
+		return vmcommon.Ok
+	} else {
+		return vmcommon.UserError
+	}
+}
+
+func (fa *freezeAccount) pending(guardian *Guardian) bool {
+	currEpoch := fa.systemEI.BlockChainHook().CurrentEpoch()
+	remaining := absDiff(currEpoch, guardian.ActivationEpoch) // any edge case here for which we should use abs?
+	return remaining < fa.guardianEnableEpochs
+}
+
+func absDiff(a, b uint32) uint32 {
+	if a < b {
+		return b - a
+	}
+	return a - b
+}
+
+func (fa *freezeAccount) addGuardian(address []byte, guardianAddress []byte, guardians Guardians) error {
+	guardian := &Guardian{
+		Address:         guardianAddress,
+		ActivationEpoch: fa.systemEI.BlockChainHook().CurrentEpoch() + fa.guardianEnableEpochs,
+	}
+
+	guardians.Data = append(guardians.Data, guardian)
+	marshalledData, err := fa.marshaller.Marshal(guardians)
+	if err != nil {
+		return err
+	}
+
+	account, err := fa.systemEI.BlockChainHook().GetUserAccount(address)
+	if err != nil {
+		return err
+	}
+
+	key := calcProtectedPrefixedKey(GuardiansKey)
+	return account.AccountDataHandler().SaveKeyValue(key, marshalledData)
```

Comment:
This is wrong. You never write from a SC directly to a UserAccount - you always write into the cache simply. you should use fa.systemEI.SetStorage. But this has to be moved to built in functions, those are the actual SC for every account.

Reply:
Thanks for clarification, will change it :pray: